### PR TITLE
dont shim GUM if unprefixed gum exists

### PR DESCRIPTION
**Pull request by fippo**
lets see if the tests pass. This needs some more discussion before actually merging.

The tests that fail (locally) are conversion tests which i'd expect to fail.

### CODE REVIEW

**Comment by fippo** [comment]
uhm... travis?!

**Comment by fippo** [comment]
https://travis-ci.org/webrtc/adapter/jobs/139182194 shows the canary output. This looks great!

**Comment by KaptenJansson** [comment]
@alvestrand FYI

**Comment by KaptenJansson** [comment]
So this means we have to disable the gUM constraints conversion tests for <=M52 versions, right?

**Comment by alvestrand** [comment]
My instinctive reaction is that this decision belongs in the Chrome shim - that is, chromeShim.shimGetUserMedia() should be the one making the decision.

So this line from src/js/chrome/getusermedia.js:

 navigator.getUserMedia = getUserMedia_;

should become

if (!navigator.getUserMedia) {
   // getuserMedia was added in 53
   navigator.getUserMedia = getUserMedia_;
}

That way, we can keep stuff isolated if we discover ways in which the new getUserMedia require shimming.

**Comment by alvestrand** [comment]
The constraints conversion tests should still pass - the old constraints are still supported, the tests explicitly target the prefixed "webkitGetUserMedia" interface, and the tests are targeted at testing the conversion functions, not how Chrome handles them.

**Comment by fippo** [comment]
the tests target webkitGetUserMedia -- but the getUserMedia call and we do not shim navigator.webkitGetUserMedia (should we?)

Just returning isn't enough, this misses the error shim. Or did you already implement that in M53 as well?

**Comment by KaptenJansson** [comment]
I do not think we should be shimming prefixed API's once they have been unprefixed.

**Comment by KaptenJansson** [comment]
This LGTM now,  @alvestrand any concerns?

**Comment by KaptenJansson** [comment]
Ping

**Comment by KaptenJansson** [comment]
Are we ready to merge this ?

File: src/js/chrome/getusermedia.js [hunk 1]
```diff
@@ -11,6 +11,9 @@ var logging = require('../utils.js').log;
 
 // Expose public methods.
 module.exports = function() {
+  if (navigator.getUserMedia) {
+    return;
+  }
```
Comment (alvestrand) [inline src/js/chrome/getusermedia.js]: Keep an eye on this one. It turns of _all_ shimming if getUserMedia is detected unprefixed, including the error shim. Would be nice if it's correct.

File: test/test.js [hunk 2]
```diff
@@ -1251,7 +1251,8 @@ test('Check getUserMedia legacy constraints converter', function(t) {
     // decision if the test should be run or not is in the Test definition
     // rather than the preferred Run test section (Webdriver).
     // FIXME: Move the decision to // Run test.
-    if (window.adapter.browserDetails.browser === 'chrome') {
+    if (window.adapter.browserDetails.browser === 'chrome'
+        && window.adapter.browserDetails.browserVersion < 53) {
```
Comment (alvestrand) [inline test/test.js]: This needs to move. The test for browserVersion < 53 should be specifically around the back-conversion of new-style constraints; the test for old constraints passing through unharmed still needs to be run.
We also need a new test - that new-style constraints are _not_ converted if browserVersion >= 53.

**Comment by alvestrand** [comment]
Sorry for being slow.

**Comment by dandv** [comment]
:+1:

I'm building an [ImageCapture takePhoto()](https://www.w3.org/TR/image-capture/#dom-imagecapture-takephoto) polyfill and calling the [Promise-style MediaDevices.gUM() API](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia) results in an odd error, given than Chrome 53 supports `navigator.getUserMedia()` unprefixed:

``` js
navigator.getUserMedia({video: true}).then(mediaStream => {
    ...
}).catch(error => {
    ...
});
```

The error is "Uncaught TypeError: Failed to execute **_'webkitGetUserMedia'**_ on 'Navigator': The callback provided as parameter 2 is not a function.".

http://jsbin.com/yerojuhudu/edit?html,css,output

**Comment by fippo** [comment]
@dandv navigator.getUserMedia doesn't do promises (which is one of the reasons it is deprecated)

**Comment by dandv** [comment]
@fippo: right, my point was that the error was surprising in that it mentioned `webkitGetUserMedia` instead of `navigator.getUserMedia`.

**Comment by fippo** [comment]
@dandv until this is merged (which is blocked by fixing the tests; i'm slow) we are overriding navigator.getUserMedia with the legacy webkit thing

**Comment by KaptenJansson** [comment]
what's the latest here?

**Comment by fippo** [comment]
i don't have the cycles to fix the tests right now :-/

**Comment by fippo** [comment]
also removing the constraints shim doesn't seem prudent, see #408

**Pull request closed without merging**